enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+With such a CRD, OVN-Kubernetes could write status to it to indicate it has configured and is using the desired network, eliminating the need for annotating a namespace.
+
+
+### IP Addressing
+
+As previously mentioned, one of the goals is to allow unique networks to have overlapping pod IP addresses. This is enabled by allowing a user to configure what CIDR to use for pod addressing when they create the network. However, this range may not overlap with the default cluster CIDR used by the default network today.
+
+Furthermore, the internal masquerade subnet and the Kubernetes service subnet will remain unique and will exist globally to serve all networks. The join subnets and transit switch subnets will overlap between networks, and are not exposed outside of the OVN topology.
+
+
+### DNS
+
+DNS lookups will happen via every pod’s access to the DNS service on the default network. CoreDNS lookups for pods will resolve to the pod’s IP on the default network. This is a limitation of the first phase of this feature and will be addressed in a future enhancement. DNS lookups for services and external entities will function correctly.
+
+
+### Services
+
+Services in Kubernetes are namespace scoped. Any creation of a service in a namespace without a unique network (using default network as primary) will only be accessible by other namespaces also using the default network as their primary network. Services created in namespaces served by unique networks, will only be accessible to namespaces connected to the unique network. Two caveats to this are:
+
+
+
+1. Host networked pods of any namespace are able to reach any service.
+2. Pods that are connected to unique networks are still able to access KAPI and DNS services that reside on the default network.
+
+With this proposal, OVN-Kubernetes will check which network is being used for this namespace, and then only enable the service there. The cluster IP of the service will only be available in the network of that service. The one caveat to this is that the service will still be reachable from any host via Cluster IP. Load balancer IP and nodeport services are also supported on unique networks. Service selectors will only be able to select endpoints for the namespace(s) attached to the network. Services that exist before the unique network is assigned to a namespace will result in OVN-Kubernetes executing a re-sync on all services in that namespace, and updating all load balancers. Keep in mind that, pods may not exist in the namespace when the namespace is assigned to a new network or the new network assignment will not be accepted by OVN-Kubernetes.
```

Comment:
@danwinship let's discuss that. I think there are these scenarios:
1. Any non-OCP node, external entity can reach the LB IP or nodeport from outside.
2. Any OCP node host networked pod could be blocked via iptables/nftables cgroup rule from reaching nodeport or load balancer IP in other networks.
3. Any OVN networked pod trying to reach a nodeport or LB IP would egress through its primary network interface (default gateway). This traffic would then hit the OVN primary logical topology that today will route this out of the node on the underlay, and it would work. We would have to add OVN ACLs to block this if we do not want to allow access.

So I guess it is possible to restrict or allow access from OCP nodes/pods, but what do we want to do?

Reply:
In general I feel like we don't need to worry about blocking people from reaching publicly-accessible services. Like, with openshift-sdn multitenant, we didn't block namespaces from reaching other namespaces' services via LB IPs/NodePorts, and nobody ever complained about that.

If a pod can reach the internet, then it can effectively reach any load balancer IP, because you can just set up a proxy somewhere else on the internet to bounce the connection back for you. So blocking pods from reaching LB IPs is only interesting if you're also blocking them from reaching the internet. But if you have EgressFirewall/AdminNetworkPolicy blocking them from reaching the internet, then you can just also block them from reaching the LB IP too.